Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp

```diff
@@ -292,22 +292,22 @@ void WebSWContextManagerConnection::fireActivateEvent(ServiceWorkerIdentifier id
     });
 }
 
-void WebSWContextManagerConnection::firePushEvent(ServiceWorkerIdentifier identifier, const std::optional<IPC::DataReference>& ipcData, CompletionHandler<void(bool)>&& callback)
+void WebSWContextManagerConnection::firePushEvent(ServiceWorkerIdentifier identifier, const std::optional<IPC::DataReference>& ipcData, std::optional<NotificationPayload>&& proposedPayload, CompletionHandler<void(bool, std::optional<NotificationPayload>&&)>&& callback)
 {
     assertIsCurrent(m_queue.get());
 
     std::optional<Vector<uint8_t>> data;
     if (ipcData)
         data = Vector<uint8_t> { ipcData->data(), ipcData->size() };
 
-    auto inQueueCallback = [queue = m_queue, callback = WTFMove(callback)](bool result) mutable {
-        queue->dispatch([result, callback = WTFMove(callback)]() mutable {
-            callback(result);
+    auto inQueueCallback = [queue = m_queue, callback = WTFMove(callback)](bool result, std::optional<NotificationPayload>&& resultPayload) mutable {
+        queue->dispatch([result, resultPayload = WTFMove(resultPayload), callback = WTFMove(callback)]() mutable {
```

Comment:
`WTFMove(resultPayload).isolatedCopy()` no? since we're passing to another thread.

Reply:
I don't understand why we have code that does `WTFMove(resultPayload).isolatedCopy()`, if moving the object to another thread is bad.

Shouldn't such cases always just be `resultPayload = resultPayload.isolatedCopy()` (assuming the return of `isolatedCopy` is automatically moved to the new lambda value?

Reply:
Many of our objects (like String) have a more optimized isolatedCopy() version when called on a r-value reference (which avoids isolated copying in cases where it is safe). Therefore, if our can WTFMove(), you should WTFMove().isolatedCopy().

Also, you should always isolatedCopy() when passing non-trivial types to another thread.

Reply:
Yah, if we deem somewhere needs an isolatedCopy(), then the WTFMove(...) part is unnecessary.

I'll check more closely on this, but I'm not convinced we need them:
It came directly off IPC and been WTFMove()'d around.

After addressing all other feedback, will look closely.

Reply:
> Yah, if we deem somewhere needs an isolatedCopy(), then the WTFMove(...) part is unnecessary.

I explained that this isn't true. `WTFMove(foo).isolatedCopy()` is more efficient with MANY of our types. See for example String:
```
    WTF_EXPORT_PRIVATE String WARN_UNUSED_RETURN isolatedCopy() const &;
    WTF_EXPORT_PRIVATE String WARN_UNUSED_RETURN isolatedCopy() &&;
```

> It came directly off IPC and been WTFMove()'d around.

We use WTFMove().isolatedCopy() these days to do this safely. It is as efficient and you can't get it wrong. It is bad practice to make assumption and just WTFMove() to another thread. We stopped doing this a while back.